Performance option: no moving convoys between CPs

MetalStormGhost opened the pull request:
Added an option to make convoys between CPs stationary, for performance reasons. This option aims to remove long routes for ground vehicles between control points, since the CPU load for long routes on DCS is pretty heavy. The convoys are actually not fully stationary, they do have a short route defined so they might find a road to drive on, but for the purposes of decreasing CPU load, this should be enough. The convoys will still be there as targets for BAI missions.

Potential known issue: convoy position might not correspond to BAI flight waypoints.

MetalStormGhost committed:
Added an option to make convoys between CPs stationary, for performance reasons. This option aims to remove long routes for ground vehicles between control points, since the CPU load for long routes on DCS is pretty heavy. The convoys are actually not fully stationary, they do have a short route defined so they might find a road to drive on, but for the purposes of decreasing CPU load, this should be enough. The convoys will still be there as targets for BAI missions.

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into perf_moving_convoys

Starfire13 commented:
I wonder whether it might be good to start them off a little further from the control point, because if they are only going to move a very short distance, they are essentially permanently under the protection umbrella of all the base defences, whereas an actual convoy would likely have moved a little distance beyond that by the time the anti-convoy flight arrives.

MetalStormGhost commented:
In the existing implementation, they are actually placed roughly halfway between the source and the destination. That will actually place them outside the protection umbrella of base defences, and will also mean they are not exactly where BAI flight waypoints are. 

Tweaking this is quite simple by modifying a multiplier used in the calculation of the position coordinates. I intend to move them a bit closer to the source CP after running some tests.

Starfire13 commented:
Sounds good. I agree that halfway is a little too far forward. Somewhere around 1/3 of the way would be nice. What do you mean by "they are not exactly where BAI flight waypoints are"? Wouldn't that mean that the AI flights wouldn't be able to locate the convoy and end up RTBing?

MetalStormGhost commented:
Yes, that's exactly what it would mean. It was also Dan Albert's primary concern with this.

Moving the convoys closer to the source CP would mitigate this somewhat.

MetalStormGhost committed:
Merge branch 'develop' of https://github.com/dcs-liberation/dcs_liberation into perf_moving_convoys

MetalStormGhost committed:
Merge branch 'perf_moving_convoys' of https://github.com/MetalStormGhost/dcs_liberation into perf_moving_convoys

MetalStormGhost committed:
Merge branch 'perf_moving_convoys' of https://github.com/MetalStormGhost/dcs_liberation into perf_moving_convoys

MetalStormGhost committed:
Moved the position of the convoy closer to the source CP when "Moving convoys between CPs" option is disabled. This will place them closer to the protection umbrella of the CP, as suggested by @Starfire13

This will also help prevent potential problems with AI flights not finding the convoy on BAI missions, since the attack waypoint is generated at the source CP.

DanAlbert commented on game/missiongenerator/convoygenerator.py:
```diff
@@ -32,17 +32,46 @@ def generate(self) -> None:
+            # Put a some distance between the convoy and its first waypoint, so it'll move a little
+            # before stopping and possibly find a road to drive on.
+            start_x = (3 * convoy.route_start.x + convoy.route_end.x) / 4
+            start_y = (3 * convoy.route_start.y + convoy.route_end.y) / 4
```
What if this is water, or a building?

I don't think this can reasonably be done without input from the campaign designer to specify more than one spawn point (https://github.com/dcs-liberation/dcs_liberation/issues/1002). Until then I'm okay with the option, but I think it still needs to spawn in the normal location instead of at an arbitrary location that might be inside a building, forest, or lake.

MetalStormGhost replied:
The waypoints are now generated from the campaign definition supply routes by utilizing Frontline. Here the direct route from Gelendzhik to Novorossiysk would have placed the convoy in the sea with the previous implementation: 
![image](https://user-images.githubusercontent.com/89945461/142253404-c3462200-6b72-4a62-b105-53b52ccb6d25.png)

The convoy is now correctly placed on land.

MetalStormGhost committed:
As suggested, changed the convoy waypoint generation when perf_moving_convoys is disabled. Now generates the waypoints from frontline segments, instead of arbitrary points, to prevent the convoy from spawning in e.g. water.

MetalStormGhost committed:
Renamed the perf_moving_convoys option because the segment based waypoint generation might generate longer paths for the convoys, depending on the campaign definition.

DanAlbert commented:
(ack that this is now waiting on me, will try to get to it soon)

DanAlbert commented on game/missiongenerator/convoygenerator.py:
```diff
@@ -32,17 +33,46 @@ def generate(self) -> None:
+                frontline.segments[convoy_segment].point_b,
+                speed=kph(40).kph,
+                move_formation=PointAction.OnRoad,
+            )
```
If they move at all, what's the point of the option? It's not going to help performance (in fact I wouldn't be surprised if off-road navigation had _worse_ performance).

MetalStormGhost replied:
I have had horrible CPU load when I've planned vehicle convoys (in manually designed missions) which have driven from Tblisi to Kutaisi (along roads). I noticed some campaigns have that exact same path between control points, such as the Multi-part Georgia campaign. I think the length of the path impacts the CPU load. My understanding is that plotting long paths for ground units with Combined Arms will lag the server, which is why it's usually disabled on multiplayer servers:
https://forums.eagle.ru/topic/279435-pathfinding-please/

This is why I believe the shorter paths imrove performance, and initial testing with a friend of mine seems to confirm this.

DanAlbert replied:
If the part that improves performance is just shorter distances, implementing https://github.com/dcs-liberation/dcs_liberation/issues/1002 would solve the performance problem and improve convoy behavior in game.

DanAlbert commented on game/missiongenerator/convoygenerator.py:
```diff
@@ -32,17 +33,46 @@ def generate(self) -> None:
+                frontline.segments[convoy_segment].point_a,
+                convoy.units,
+                convoy.player_owned,
+            )
```
This could be nowhere remotely near the BAI target area. I don't think it's worth fighting all the edge cases in this when we could instead just spawn at the origin point. The convoys just barely leave their origin in the span of a mission anyway, so it's a more accurate representation to have them there than 1/3 of the way between bases anyway.

I don't think the code here needs to be any more complicated than an `if` around the existing `add_waypoint`.

MetalStormGhost replied:
Implemented the suggested change.

MetalStormGhost committed:
Implemented the suggested change by DanAlbert:

This could be nowhere remotely near the BAI target area. I don't think it's worth fighting all the edge cases in this when we could instead just spawn at the origin point. The convoys just barely leave their origin in the span of a mission anyway, so it's a more accurate representation to have them there than 1/3 of the way between bases anyway.

I don't think the code here needs to be any more complicated than an if around the existing add_waypoint.

DanAlbert commented on game/settings/settings.py:
```diff
@@ -407,6 +407,12 @@ class Settings:
         default=True,
     )
+    perf_moving_convoys: bool = boolean_option(
+        "Convoys drive the full distance between control points",
```
The name and the description do not match. This should be named `convoys_travel_full_distance` if that's the behavior it controls.

MetalStormGhost replied:
Renamed as suggested.

DanAlbert commented on game/missiongenerator/convoygenerator.py:
```diff
@@ -38,11 +39,28 @@ def generate_convoy(self, convoy: Convoy) -> VehicleGroup:
+            # so the convoy spawns at the origin CP. This allows the convoy to be
+            # targeted by BAI flights and starts it within the protection umbrella of the CP.
+            group.add_waypoint(
+                frontline.segments[0].point_b,
```
There's no guarantee that this is any shorter than `convoy.route_end`, is there?

Also, refactor. You've duplicated everything but this argument.

```python
if self.game.settings.perf_moving_convoys:
    end_point = convoy.route_end
else:
    end_point = frontline.segments[0].point_b
```

MetalStormGhost replied:
Done. No guarantees of course, but I haven't seen a campaign where the supply routes would only contain one waypoint.

MetalStormGhost committed:
Renamed the option to convoys_travel_full_distance as suggested by ColonelPanic. Also removed the unnecessarily duplicated call to group.add_waypoint().
